src/scheduler/Scheduler.cpp

```diff
@@ -303,6 +309,9 @@ faabric::util::SchedulingDecision Scheduler::makeSchedulingDecision(
 
         // Work out how many we can handle locally
         int slots = thisHostResources.slots();
+        if (topologyHint == faabric::util::SchedulingTopologyHint::UNDERFULL) {
```

Comment:
This is all the logic required to implement the `UNDERFULL` topology hint. I could move it to a separate PR but it is literally just this and the tests.